src/module.c

```diff
@@ -1109,213 +1080,667 @@ int RM_CreateSubcommand(RedisModuleCommand *parent, const char *name, RedisModul
+ *
+ * Affects the output of `COMMAND`, `COMMAND INFO` and `COMMAND DOCS`, Cluster,
+ * ACL and is used to filter commands with the wrong number of arguments before
+ * the call reaches the module code.
+ *
+ * This function can be called after creating a command using RM_CreateCommand
+ * and fetching the command pointer using RM_GetCommand. The information can
+ * only be set once for each command and has the following structure:
+ *
+ *     typedef struct RedisModuleCommandInfo {
+ *         int version;
+ *         const char *summary;
+ *         const char *complexity;
+ *         const char *since;
+ *         RedisModuleCommandHistoryEntry *history;
+ *         const char **tips;
+ *         int arity;
+ *         RedisModuleCommandKeySpec *key_specs;
+ *         RedisModuleCommandArg *args;
+ *     } RedisModuleCommandInfo;
+ *
+ * All fields are optional except `version`. Explanation of the fields:
+ *
+ * - `version`: Set this to REDISMODULE_COMMAND_INFO_VERSION. It makes the
+ *   module backward and forward compatible with older and newer Redis versions
+ *   and allows Redis to figure out which fields are defined, even if the module
+ *   is compiled with a `redismodule.h` copied from another Redis version.
+ *
+ * - `summary`: A short description of the command (optional).
+ *
+ * - `complexity`: Complexity description (optional).
+ *
+ * - `since`: The version where the command was introduced (optional).
+ *   Note: The version specified should be the module's, not Redis version.
+ *
+ * - `history`: An array of RedisModuleCommandHistoryEntry (optional), which is
+ *   a struct with the following fields:
+ *
+ *         const char *since;
+ *         const char *changes;
+ *
+ *     `since` is a version string and `changes` is a string describing the
+ *     changes. The array is terminated by a zeroed entry, i.e. an entry with
+ *     both strings set to NULL.
+ *
+ * - `tips`: A NULL-terminated array of strings that are meant to be tips for
+ *   clients and proxies regarding this command (optional). See
+ *   https://redis.io/topics/command-tips.
+ *
+ * - `arity`: Number of arguments, including the command name itself. A positive
+ *   number specifies an exact number of arguments and a negative number
+ *   specifies a minimum number of arguments, so use -N to say >= N.
+ *
+ * - `key_specs`: An array of RedisModuleCommandKeySpec, terminated by an
+ *   element memset to zero. This is a scheme that tries to describe the
+ *   positions of key arguments better than the old RM_CreateCommand arguments
+ *   `firstkey`, `lastkey`, `keystep` and is needed if those three are not
+ *   enough to describe the key positions. There are two steps to retrieve key
+ *   positions: *begin search* (BS) in which index should find the first key and
+ *   *find keys* (FK) which, relative to the output of BS, describes how can we
+ *   will which arguments are keys. Additionally, there are key specific flags.
```

Comment:
The legacy triple is updated when keyspecs is set and a default key-spec is creates from the old triple, so why do they need to set both?

Brw, isn't it high prio to make cluster and getkeys use the new key-specs?

Reply:
by both, i didn't mean the initial triple, and key specs.
although that too might be true.
i.e. that modules need to use the initial triple if they want to be compatible with old redis versions (and still show something in COMMAND command).

what i meant about both is both key-specs (i consider the initial triple to be part of the key-spsc), and the `getkeys-api` mechanism.
i.e. some mechanisms in redis user only the later, and some only the first.
and also need to keep in mind the fact that some complicated commands may not be possible to define with key-specs, in which case they need to use the invalid spec and implement the `getkeys-api`.

so we somehow need to document all of that.
maybe we don't want to document which mechanism in redis uses which, but we do want to mention that both need to be handled (i.e. both specs, and getkeys-api)

Reply:
Yes, this is a mess. It's too much to require both from a module.

When documenting this, I'd like to at least mention when/why the keypos-api is needed, but we shouldn't mention too much because we want to change it...

Ideally, the getkeys-api is only needed for UNKNOWN keyspecs and movable keys without keyspec. Can we improve this in a separate PR?

Reply:
i think i'd rather finish this discussion in this PR, so we can look at the outcome and decide if we're happy with it.
p.s. we did decide that we're surely not including this in RC1.. don't wanna make any rush decisions.. we need time to process this and be sure we have a solid API.